packages/vm/src/instance.rs

```diff
@@ -84,102 +78,135 @@ where
+        // Returns an iterator ID.
         #[cfg(feature = "iterator")]
-        import_obj.extend(imports! {
-            "env" => {
-                // Creates an iterator that will go from start to end.
-                // If start_ptr == 0, the start is unbounded.
-                // If end_ptr == 0, the end is unbounded.
-                // Order is defined in cosmwasm_std::Order and may be 1 (ascending) or 2 (descending). All other values result in an error.
-                // Ownership of both start and end pointer is not transferred to the host.
-                // Returns an iterator ID.
-                "db_scan" => Func::new(move |ctx: &mut Ctx, start_ptr: u32, end_ptr: u32, order: i32| -> VmResult<u32> {
-                    do_scan::<S, Q>(ctx, start_ptr, end_ptr, order)
-                }),
-                // Get next element of iterator with ID `iterator_id`.
-                // Creates a region containing both key and value and returns its address.
-                // Ownership of the result region is transferred to the contract.
-                // The KV region uses the format value || key || keylen, where keylen is a fixed size big endian u32 value.
-                // An empty key (i.e. KV region ends with \0\0\0\0) means no more element, no matter what the value is.
-                "db_next" => Func::new(move |ctx: &mut Ctx, iterator_id: u32| -> VmResult<u32> {
-                    do_next::<S, Q>(ctx, iterator_id)
-                }),
-            },
-        });
+        env_imports.insert(
+            "db_scan",
+            Function::new_native_with_env(store, env.clone(), native_db_scan),
+        );
+
+        // Get next element of iterator with ID `iterator_id`.
+        // Creates a region containing both key and value and returns its address.
+        // Ownership of the result region is transferred to the contract.
+        // The KV region uses the format value || key || keylen, where keylen is a fixed size big endian u32 value.
+        // An empty key (i.e. KV region ends with \0\0\0\0) means no more element, no matter what the value is.
+        #[cfg(feature = "iterator")]
+        env_imports.insert(
+            "db_next",
+            Function::new_native_with_env(store, env.clone(), native_db_next),
+        );
+
+        import_obj.register("env", env_imports);
 
-        let mut wasmer_instance =
-            Box::from(module.instantiate(&import_obj).map_err(|original| {
+        let wasmer_instance = Box::from(WasmerInstance::new(&module, &import_obj).map_err(
+            |original| {
                 VmError::instantiation_err(format!("Error instantiating module: {:?}", original))
-            })?);
+            },
+        )?);
 
-        set_gas_left(wasmer_instance.context_mut(), gas_limit);
-        get_gas_state_mut::<S, Q>(wasmer_instance.context_mut()).set_gas_limit(gas_limit);
+        set_gas_left(&env, gas_limit);
```

Comment:
Okay, gas limit is set on each instantiation.
The memory limit can only be set on compile/creating the module, right?

Reply:
A Module is Artifact + Store. We don't really cache Modules, we cache Artifacts. Once we restore and Artifact and add a Store to it, we get a Module with a memory limit set.

So yes, the limit is in the module but it is set every time we load from cache.